<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>Population genetic model</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SR meiotic drive</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Population genetic model</h1>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2019-08-14
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>cost_of_SR_Dpseudo/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.3.0). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20190319code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20190319)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20190319code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20190319)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomlukeholmancostofSRDpseudotreee5f7926c0ebfdf8a7498e250b093b90dd0ab3c88targetblanke5f7926a"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/lukeholman/cost_of_SR_Dpseudo/tree/e5f7926c0ebfdf8a7498e250b093b90dd0ab3c88" target="_blank">e5f7926</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomlukeholmancostofSRDpseudotreee5f7926c0ebfdf8a7498e250b093b90dd0ab3c88targetblanke5f7926a" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility. The version displayed above was the version of the Git repository at the time these results were generated. <br><br> Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    output/.DS_Store

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the R Markdown and HTML files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view them.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/lukeholman/cost_of_SR_Dpseudo/e5f7926c0ebfdf8a7498e250b093b90dd0ab3c88/docs/population_genetic_model.html" target="_blank">e5f7926</a>
</td>
<td>
lukeholman
</td>
<td>
2019-08-14
</td>
<td>
working nicely
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/lukeholman/cost_of_SR_Dpseudo/24d4c9c8efd48d58a41fc10fa2e4a20c565d1768/docs/population_genetic_model.html" target="_blank">24d4c9c</a>
</td>
<td>
lukeholman
</td>
<td>
2019-07-02
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/lukeholman/cost_of_SR_Dpseudo/blob/0479dbb377a67982268f83264c29ac6e186107f3/analysis/population_genetic_model.Rmd" target="_blank">0479dbb</a>
</td>
<td>
lukeholman
</td>
<td>
2019-07-02
</td>
<td>
re-run model
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/lukeholman/cost_of_SR_Dpseudo/93aec6e1d8dad8307c93f829030a82f004c39245/docs/population_genetic_model.html" target="_blank">93aec6e</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/lukeholman/cost_of_SR_Dpseudo/blob/01d7dbbcb24a7b6a4c779d3af9f5cd17545b4ad2/analysis/population_genetic_model.Rmd" target="_blank">01d7dbb</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
<td>
First complete commit
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/lukeholman/cost_of_SR_Dpseudo/8c6e0c3ffde1b51ff09785d32ca823a3c0df961e/docs/population_genetic_model.html" target="_blank">8c6e0c3</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/lukeholman/cost_of_SR_Dpseudo/blob/ffdc5d44370af9cba1f49ab7045fd537d8a2191c/analysis/population_genetic_model.Rmd" target="_blank">ffdc5d4</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
<td>
First complete commit
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/lukeholman/cost_of_SR_Dpseudo/ffdc5d44370af9cba1f49ab7045fd537d8a2191c/docs/population_genetic_model.html" target="_blank">ffdc5d4</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
<td>
First complete commit
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(purrr)
library(dplyr)
library(future)
library(future.apply)
library(kableExtra)
library(ggplot2)
library(showtext)
font_add_google(name = &quot;Lato&quot;, family = &quot;Lato&quot;, regular.wt = 400, bold.wt = 700)
showtext_auto()
options(stringsAsFactors = FALSE)</code></pre>
<div id="enter-the-real-world-data-on-polyandry-and-sr-frequency" class="section level2">
<h2>Enter the real-world data on polyandry and SR frequency</h2>
<pre class="r"><code>real_world_data &lt;- data.frame(
  real_freq_SR   = c(10.2, 13.5, 11.5, 11.0, 7.5, 3.5, 0),
  freq_polyandry = c(0.58, 0.52, 0.73, 0.59, 0.73, 0.73, 0.92)
)

# The average of P1 and P2 in from Giraldo-Perez et al. 2016, https://onlinelibrary.wiley.com/doi/pdf/10.1111/jeb.12792
average_paternity_SR_males &lt;- 0.2105

# The estimated frequency of SR male progeny who inherit SR, REFERENCE?
real_world_k &lt;- 0.96</code></pre>
</div>
<div id="function-to-make-the-mating-table" class="section level2">
<h2>Function to make the mating table</h2>
<p>The “mating table” is list of all the possible mating types (e.g. ST/SR mother with a SR/Y father, etc etc), which lists the possible offspring genotypes and their associated frequencies. The precise values in the mating table are a function of <span class="math inline">\(k\)</span> (i.e. the strength of segregation distortion) and expected paternity of SR males when the female is double-mated to an ST and a SR male. We assume that the meiosis is Mendelian except in SR males, and that all crosses except those with a SR father produce a 50:50 sex ratio.</p>
<pre class="r"><code>make_mating_table &lt;- function(k, paternity_of_SR_males, mating_types){

  offspring &lt;- vector(mode = &quot;list&quot;, nrow(mating_types))
  
  make_offspring &lt;- function(row, type, zygote_freq){
    data.frame(mother = mating_types$mother[row],
               father1 = mating_types$father1[row],
               father2 = mating_types$father2[row],
               type, 
               zygote_freq, 
               stringsAsFactors = FALSE)
  }
  
  g0 &lt;- c(&quot;STST_female&quot;, &quot;STSR_female&quot;,      # possible offspring types
          &quot;ST_male&quot;, &quot;SR_male&quot;)
  g1 &lt;- c(&quot;STSR_female&quot;, &quot;SRSR_female&quot;,
          &quot;ST_male&quot;, &quot;SR_male&quot;)

  p1 &lt;- c(0.5 *  (1 + k), 0.5 * (1 - k))
  p2 &lt;- c(0.25 * (1 + k), 0.25 * (1 + k),
          0.25 * (1 - k), 0.25 * (1 - k))  
  
  p3 &lt;- c(0.5 * (1 - paternity_of_SR_males), # ST male wins, he transmits his ST-X with 50% chance
          0.5 * paternity_of_SR_males * (1 + k), # SR male wins with probability sc, he transmits his SR-X 0.5*(1+k) chance
          1 - 0.5 * (1 - paternity_of_SR_males) - 0.5 * paternity_of_SR_males * (1 + k)) # sons created by the remaining two possibilities

  p4 &lt;- c(0.25 * (1 - paternity_of_SR_males), # ST male wins, he transmits his ST-X with 50% chance, same X in female meiosis
          0.25 * (1 - paternity_of_SR_males) + 0.25 * paternity_of_SR_males * (1 + k), # Either male wins and transmits the other X than the female does
          0.25 * paternity_of_SR_males * (1 + k), # SR male wins, he transmits his SR-X 0.5*(1+k) chance, same X in female meiosis
          0, 0)
  p4[4:5] &lt;- (1 - sum(p4[1:3])) / 2 # sons created by the remaining possibilities

  rep25 &lt;- rep(0.25, 4)
  rep5  &lt;- rep(0.5, 2)
  
  # Single mating types
  offspring[[1]]  &lt;- make_offspring(1, c(&quot;STST_female&quot;, &quot;ST_male&quot;), rep5)
  offspring[[2]]  &lt;- make_offspring(2, g0, rep25)
  offspring[[3]]  &lt;- make_offspring(3, c(&quot;STSR_female&quot;, &quot;SR_male&quot;), rep5)
  offspring[[4]]  &lt;- make_offspring(4, c(&quot;STSR_female&quot;, &quot;ST_male&quot;), p1)
  offspring[[5]]  &lt;- make_offspring(5, g1, p2)
  offspring[[6]]  &lt;- make_offspring(6, c(&quot;SRSR_female&quot;, &quot;SR_male&quot;), p1)
  
  # Double mating types - two ST males
  offspring[[7]]  &lt;- make_offspring(7, c(&quot;STST_female&quot;, &quot;ST_male&quot;), rep5)
  offspring[[8]]  &lt;- make_offspring(8, g0, rep25)
  offspring[[9]]  &lt;- make_offspring(9, c(&quot;STSR_female&quot;, &quot;SR_male&quot;), rep5)
  
  # Double mating types - ST and SR males
  offspring[[10]]  &lt;- make_offspring(10, c(&quot;STST_female&quot;, &quot;STSR_female&quot;, &quot;ST_male&quot;), p3)
  offspring[[11]]  &lt;- make_offspring(11, c(&quot;STST_female&quot;, 
                                           &quot;STSR_female&quot;, 
                                           &quot;SRSR_female&quot;, 
                                           &quot;ST_male&quot;, 
                                           &quot;SR_male&quot;), p4)
  offspring[[12]]  &lt;- make_offspring(12, c(&quot;STSR_female&quot;, &quot;SRSR_female&quot;, &quot;SR_male&quot;), p3)


  # Double mating types - two SR males
  offspring[[13]]  &lt;- make_offspring(13, c(&quot;STSR_female&quot;, &quot;ST_male&quot;), p1)
  offspring[[14]]  &lt;- make_offspring(14, g1, p2)
  offspring[[15]]  &lt;- make_offspring(15, c(&quot;SRSR_female&quot;, &quot;SR_male&quot;), p1)
  
  output &lt;- do.call(&quot;rbind&quot;, offspring)
  names(offspring) &lt;- paste(mating_types[,1], mating_types[,2], sep = &quot; x &quot;)
  output
}</code></pre>
<div id="an-example-of-a-mating-type-table" class="section level3">
<h3>An example of a mating type table</h3>
<p>An example of the output of <code>make_mating_table()</code>, under the assumptions that:</p>
<ol style="list-style-type: decimal">
<li>there is strong but incomplete segregation distortion (<code>k = 0.9</code>),</li>
<li>SR males are slightly less than half as effective in sperm competition (<code>paternity_of_SR_males = 0.23</code>).</li>
</ol>
<pre class="r"><code>make_mating_types &lt;- function(){
  expand.grid(mother = c(&quot;STST_female&quot;,
                         &quot;STSR_female&quot;,
                         &quot;SRSR_female&quot;),
              father1 = c(&quot;ST_male&quot;,
                          &quot;SR_male&quot;),
              father2 = c(&quot;none&quot;,
                          &quot;ST_male&quot;,
                          &quot;SR_male&quot;),
              stringsAsFactors = FALSE) %&gt;%
    mutate(sorted_paste = map_chr(1:n(), function(i) paste0(sort(c(mother[i], father1[i], father2[i])), collapse = &quot;&quot;))) %&gt;%
    distinct(sorted_paste, .keep_all = TRUE) %&gt;% select(-sorted_paste)
}

mating_types &lt;- make_mating_types()

make_mating_table(k = 0.9, paternity_of_SR_males = 0.23, mating_types) %&gt;%
  rename(zygote_type = type) %&gt;%
  kable() %&gt;%
  kable_styling() %&gt;%
  scroll_box(height = &quot;500px&quot;)</code></pre>
<div style="border: 1px solid #ddd; padding: 5px; overflow-y: scroll; height:500px; ">
<table class="table" style="margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;">
mother
</th>
<th style="text-align:left;">
father1
</th>
<th style="text-align:left;">
father2
</th>
<th style="text-align:left;">
zygote_type
</th>
<th style="text-align:right;">
zygote_freq
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.95000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.05000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.47500
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:right;">
0.47500
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.02500
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.02500
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:right;">
0.95000
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
none
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.05000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.25000
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.50000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:right;">
0.38500
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.21850
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.39650
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:right;">
0.19250
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.30175
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:right;">
0.10925
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.19825
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.19825
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.38500
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:right;">
0.21850
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.39650
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.95000
</td>
</tr>
<tr>
<td style="text-align:left;">
STST_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.05000
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:right;">
0.47500
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:right;">
0.47500
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
ST_male
</td>
<td style="text-align:right;">
0.02500
</td>
</tr>
<tr>
<td style="text-align:left;">
STSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.02500
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:right;">
0.95000
</td>
</tr>
<tr>
<td style="text-align:left;">
SRSR_female
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:left;">
SR_male
</td>
<td style="text-align:right;">
0.05000
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="function-to-determine-the-mating-type-frequencies-for-the-population" class="section level2">
<h2>Function to determine the mating type frequencies for the population</h2>
<p>Given a set of genotype frequencies, it is simple to calculate the frequencies of each mating type. First, we implement selection, such that each genotype is represented in the mating types according to the product of its frequency and its fitness. Second, we determine the frequency of cross between male genotype <span class="math inline">\(i\)</span> and female genotype <span class="math inline">\(j\)</span> as <span class="math inline">\(i\times j\)</span>.</p>
<pre class="r"><code>find_mating_type_frequencies &lt;- function(pop, mating_types, freq_polyandry){

  # Implement viability selection
  pop$prop &lt;- pop$prop * pop$fitness
  pop$prop &lt;- pop$prop / sum(pop$prop)
  
  single_maters &lt;- 1:6
  double_maters &lt;- 7:15
  
  # Mating type freqs for the single-mating females is freq_single_maters * freq_females * freq_males, post selection
  mating_types$mating_freq[single_maters] &lt;- (1 - freq_polyandry) * 
    pop$prop[match(mating_types$mother[single_maters], pop$type)] * 
    pop$prop[match(mating_types$father1[single_maters], pop$type)]
  
  # Mating type freqs for the single-mating females is freq_double_maters * freq_females * freq_father1 * freq_father2, post selection
  # NB the ST + SR double matings are doubled, since they can occur with either male as father1
  mating_types$mating_freq[double_maters] &lt;- freq_polyandry * 
    pop$prop[match(mating_types$mother[double_maters], pop$type)] * 
    pop$prop[match(mating_types$father1[double_maters], pop$type)] *
    pop$prop[match(mating_types$father2[double_maters], pop$type)] 
  
  mating_types$mating_freq[10:12] &lt;- 2 * mating_types$mating_freq[10:12]
  mating_types$mating_freq &lt;- mating_types$mating_freq / sum(mating_types$mating_freq) # ensure the mating freqs sum to one

  mating_types
}</code></pre>
</div>
<div id="define-the-main-simulation-function" class="section level2">
<h2>Define the main simulation function</h2>
<p>This function iterates over generations, implementing selection and reproduction each time, until either A) the SR alelle fixes, B) the SR alelle goes extinct, or C) the generation timer expires.</p>
<pre class="r"><code># Helper function to calculate the frequency of the SR allele
calc_prop_SR &lt;- function(pop){
  sum(c(pop$prop[c(1, 4)], 2 * pop$prop[2])) / (sum(c(2 * pop$prop[c(2,4,5)], pop$prop[c(1,3)])))
}

run_simulation &lt;- function(generations,
                           k,
                           paternity_of_SR_males,
                           freq_polyandry,
                           w_STSR_female,
                           w_SRSR_female,
                           w_SR_male,
                           initial_freq_SR,
                           mating_types){
  
  # Make the initial population
  pop &lt;- data.frame(
    type = c(&quot;STST_female&quot;,
             &quot;STSR_female&quot;,
             &quot;SRSR_female&quot;,
             &quot;ST_male&quot;,
             &quot;SR_male&quot;),
    prop = c(0.5 * (1 - initial_freq_SR) ^ 2,
             (1 - initial_freq_SR) * initial_freq_SR,
             0.5 * initial_freq_SR ^ 2,
             0.5 * (1 - initial_freq_SR),
             0.5 * initial_freq_SR), 
    fitness = c(1, w_STSR_female, w_SRSR_female, 1, w_SR_male),
    stringsAsFactors = FALSE) %&gt;% 
    arrange(type) 

  # Make the mating table for this specific parameter space
  mating_table &lt;- make_mating_table(k, paternity_of_SR_males, mating_types)
  
  # Iterate over generations
  for(i in 1:generations){
    
    # Find the mating type frequencies, find the offspring frequencies, and renormalise the frequencies to sum to one
    prop_col &lt;- mating_table %&gt;%
      left_join(find_mating_type_frequencies(pop, mating_types, freq_polyandry), 
                by = c(&quot;mother&quot;, &quot;father1&quot;, &quot;father2&quot;)) %&gt;%
      mutate(offspring_freq = mating_freq * zygote_freq) %&gt;%
      group_by(type) %&gt;%                         # sorts by type?
      summarise(prop = sum(offspring_freq)) %&gt;%
      mutate(prop = prop / sum(prop)) %&gt;%
      pull(prop)
    
    # Calculate the frequency of SR. 
    # Quit early if SR fixed/extinct-ish
    prop_SR &lt;- calc_prop_SR(pop %&gt;% mutate(prop = prop_col))
    # if(is.na(prop_SR)) print(pop)   # FOR DEBUGGING
    pop &lt;- pop %&gt;% mutate(prop = prop_col)
        
    if(prop_SR &gt; 0.99) return(pop) 
    if(prop_SR &lt; 0.0001) return(pop)
  }
  pop
}</code></pre>
</div>
<div id="define-a-helper-function-to-run-the-main-simulation-on-a-data-frame-of-parameters" class="section level2">
<h2>Define a helper function to run the main simulation on a data frame of parameters</h2>
<p>A helper function that runs <code>run_simulation()</code> on a data frame of parameter values.</p>
<pre class="r"><code>run_many_simulations &lt;- function(parameters){
  
  mating_types &lt;- make_mating_types() %&gt;% mutate(mating_freq = NA)
  
  # Carefully pass all the parameters to run_simulation()
  run_one_simulation &lt;- function(row, parameters, mating_types){
    
    genotypes &lt;- run_simulation(
      generations = parameters$generations[row],
      k = parameters$k[row],
      paternity_of_SR_males = parameters$paternity_of_SR_males[row],
      freq_polyandry = parameters$freq_polyandry[row],
      w_STSR_female = parameters$w_STSR_female[row],
      w_SRSR_female = parameters$w_SRSR_female[row],
      w_SR_male = parameters$w_SR_male[row],
      initial_freq_SR = parameters$initial_freq_SR[row],
      mating_types = mating_types
    )
    
    output &lt;- data.frame(parameters[row, ]) %&gt;% as_tibble()
    output$genotypes &lt;- list(genotypes %&gt;% select(-fitness))
    output
  }
  
  # Loop over all the different parameter spaces
  lapply(1:nrow(parameters), run_one_simulation, parameters = parameters, mating_types = mating_types) %&gt;%
    do.call(&quot;rbind&quot;, .) %&gt;% 
    mutate(prop_SR = map_dbl(genotypes, calc_prop_SR))
}</code></pre>
</div>
<div id="define-the-parameter-space-to-investigate" class="section level2">
<h2>Define the parameter space to investigate</h2>
<p>Used to define the parameter spaces that are plotted in Figure 2.</p>
<pre class="r"><code>resolution &lt;- 101
gen &lt;- 100
initial_freq &lt;- 0.1

parameters &lt;- rbind(
  
   expand.grid(
    generations = gen,
    k = real_world_k,
    paternity_of_SR_males = seq(0, 0.5, length = resolution),
    freq_polyandry = seq(0, 1, length = resolution),
    w_STSR_female = 0.88, # Fitness from our Table of results
    w_SRSR_female = 0.42, # Fitness from our Table of results
    w_SR_male = 1,
    initial_freq_SR = initial_freq,
    facet = &quot;Recessive cost in females&quot;)
  
) %&gt;% as_tibble()</code></pre>
</div>
<div id="run-the-simulation" class="section level2">
<h2>Run the simulation</h2>
<pre class="r"><code>cores &lt;- 7

run_all &lt;- function(){
  if(!file.exists(&quot;data/simulation_output.rds&quot;)){ # if there is no data yet....
    
    # Divide the job into 100 equal-ish chunks so that we can see the % completion as it proceeds
    split_parameters &lt;- split(parameters, sample(1:100, nrow(parameters), replace = TRUE))
    n_chunks &lt;- length(split_parameters)
    for(i in 1:n_chunks){
      print(paste(i, &quot;out of&quot;, n_chunks))
      plan(&quot;multicore&quot;)
      chunk_size &lt;- floor(nrow(split_parameters[[i]]) / cores)
      simulation_output &lt;- split(split_parameters[[i]], 
                                 c(rep(1:(cores-1), each = chunk_size), 
                                   rep(cores, nrow(split_parameters[[i]]) - chunk_size*(cores-1)))) %&gt;%
        future_lapply(run_many_simulations) %&gt;% do.call(&quot;rbind&quot;, .)
      saveRDS(simulation_output, paste(&quot;data/simulation_output_&quot;, i, &quot;.rds&quot;, sep = &quot;&quot;))
    }
    
  } else { # if there is already some data...
    
    simulation_output &lt;- readRDS(&quot;data/simulation_output.rds&quot;)
    
    if(nrow(simulation_output) != nrow(parameters)){
      pasted_done &lt;- apply(simulation_output[, names(simulation_output) %in% names(parameters)], 1, paste0, collapse = &quot;_&quot;)
      pasted_to_do &lt;- apply(parameters, 1, paste0, collapse = &quot;_&quot;)
      trimmed_parameters &lt;- parameters[!(pasted_to_do %in% pasted_done), ]
      
      # Divide the job into 100 equal-ish chunks so that we can see the % completion as it proceeds
      split_parameters &lt;- split(trimmed_parameters, sample(1:10, nrow(parameters), replace = TRUE))
      n_chunks &lt;- length(split_parameters)
      for(i in 1:length(split_parameters)){
        print(paste(i, &quot;out of&quot;, n_chunks, sep = &quot;&quot;))
        plan(&quot;multicore&quot;)
        chunk_size &lt;- floor(nrow(split_parameters[[i]]) / cores)
        simulation_output &lt;- split(split_parameters[[i]], 
                                   c(rep(1:(cores-1), each = chunk_size), 
                                     rep(cores, nrow(split_parameters[[i]]) - chunk_size*(cores-1)))) %&gt;%
          future_lapply(run_many_simulations) %&gt;% do.call(&quot;rbind&quot;, .)
        saveRDS(simulation_output, paste(&quot;data/simulation_output_surplus_&quot;, i, &quot;.rds&quot;, sep = &quot;&quot;))
      }
    }
  }
  
  simulation_output &lt;- list.files(path = &quot;data&quot;, pattern = &quot;simulation_output_&quot;, full.names = TRUE) %&gt;%
    lapply(readRDS) %&gt;% do.call(&quot;rbind&quot;, .)
  saveRDS(simulation_output, &quot;data/simulation_output.rds&quot;)
}

run_all()
simulation_output &lt;- readRDS(&quot;data/simulation_output.rds&quot;)
  
# use to delete and re-run:    unlink(&quot;data/simulation_output.rds&quot;)
# use to delete individual files:     unlink(list.files(path = &quot;data&quot;, pattern = &quot;simulation_output_&quot;, full.names = TRUE))</code></pre>
</div>
<div id="make-figure-2" class="section level2">
<h2>Make Figure 2</h2>
<pre class="r"><code>figure_2 &lt;- simulation_output %&gt;% 
  filter(w_SR_male == 1) %&gt;%
  mutate(k = paste(&quot;Strength of drive:&quot;, k)) %&gt;%
  ggplot(aes(100 * freq_polyandry, 100 * paternity_of_SR_males, fill = 100 * prop_SR)) +
  geom_blank() +
  geom_raster() + 
  stat_contour(aes(z = 100 * prop_SR), colour = &quot;grey10&quot;, binwidth = 10) +
  geom_hline(yintercept = 50, colour = &quot;black&quot;, linetype = 3) +
  labs(x = &quot;% females that mate twice&quot;, 
       y = &quot;% paternity of SR males in a double mating&quot;) +
  annotate(&quot;pointrange&quot;, x = 100 * real_world_data$freq_polyandry, y = 21.05, 
           pch = 21, alpha = 0.7,
           fill = &quot;tomato&quot;, size = .3, ymin = 20, ymax = 20, colour = &quot;grey10&quot;) +
  scale_fill_distiller(palette = &quot;YlGnBu&quot;, direction =  1, name = &quot;% SR at equilibrium&quot;) + #, limits = c(0,100)) +
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0)) +   # , labels = c(0, 0.25, 0.5, 0.75, 1)
  theme(panel.border = element_rect(fill = NA, colour = &quot;black&quot;, size = .8),
        text = element_text(family = &quot;Lato&quot;, size = 12),
        strip.background = element_rect(colour = &quot;black&quot;, fill = &quot;grey90&quot;, size = .8),
        legend.position = &quot;top&quot;)

figure_2 %&gt;% ggsave(filename = &quot;figures/figure_2.pdf&quot;, height = 4.9, width = 4.5)
figure_2</code></pre>
<p><img src="figure/population_genetic_model.Rmd/unnamed-chunk-10-1.png" width="432" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-10-1">
Past versions of unnamed-chunk-10-1.png
</button>
</p>
<div id="fig-unnamed-chunk-10-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lukeholman/cost_of_SR_Dpseudo/blob/ffdc5d44370af9cba1f49ab7045fd537d8a2191c/docs/figure/population_genetic_model.Rmd/unnamed-chunk-10-1.png" target="_blank">ffdc5d4</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><br></br> <strong>Figure 2</strong>: Predicted equilibrium frequency of the <em>SR</em> allele, calculated from the population genetic model. The model shows that <em>SR</em> is predicted to reach a lower equilibrium frequency when a high proportion of females mate multiply (<em>x</em>-axis), and when <em>SR</em> males are inferior sperm competitors to <em>ST</em> males (y-axis). These two predictors interact, because sperm competition becomes more selectively important as polyandry becomes more common. The seven red points illustrate the range of female mating frequencies observed observed across seven North American populations, and their position on the y-axis (0.21) uses data on <em>SR</em> male siring success from Giraldo-Perez et al. 2016. The figure further assumes that <em>SR</em> males pass on the <em>SR</em> chromosome to 96% of their offspring, and that <em>ST</em> and <em>SR</em> males have equal survival and mating success.</p>
</div>
<div id="make-figure-4" class="section level2">
<h2>Make Figure 4</h2>
<pre class="r"><code>parameters &lt;- rbind(
  
   expand.grid(
    generations = gen,
    k = real_world_k, 
    paternity_of_SR_males = average_paternity_SR_males, 
    freq_polyandry = real_world_data$freq_polyandry,
    w_STSR_female = 0.88, # Fitness from our Table of results
    w_SRSR_female = 0.42, # Fitness from our Table of results
    w_SR_male = c(0.8, 0.9, 1),
    initial_freq_SR = initial_freq,
    facet = &quot;Recessive cost in females&quot;)
  
) %&gt;% as_tibble()

observed_vs_expected_data &lt;- 
  run_many_simulations(parameters) %&gt;%
  left_join(real_world_data, by = &quot;freq_polyandry&quot;) %&gt;%
  rename(predicted_p_SR = prop_SR)

figure_3 &lt;- observed_vs_expected_data %&gt;%
  mutate(w_SR_male = paste(&quot;Fitness of SR males:&quot;, w_SR_male)) %&gt;%
  ggplot(aes(real_freq_SR, 100 * predicted_p_SR, fill = 100*freq_polyandry)) +
  geom_abline(slope = 1, intercept=0, linetype=2) + 
  geom_point(pch=21, size = 3) + 
  scale_x_continuous(limits =c(0, 31)) +
  scale_y_continuous(limits =c(0, 31)) +
  scale_fill_distiller(palette = &quot;Reds&quot;, direction = 1, name = &quot;% polyandry&quot;) + 
  labs(x=&quot;Observed frequency of SR (%)&quot;,
       y=&quot;Predicted frequency of SR (%)&quot;) + 
  facet_wrap(~w_SR_male) + 
  theme_bw() + 
  theme(strip.background = element_blank(),
        text = element_text(family = &quot;Lato&quot;, size = 12)
)

figure_3 %&gt;% ggsave(filename = &quot;figures/figure_3.pdf&quot;, height = 3.3, width = 8)
figure_3</code></pre>
<p><img src="figure/population_genetic_model.Rmd/unnamed-chunk-11-1.png" width="768" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-11-1">
Past versions of unnamed-chunk-11-1.png
</button>
</p>
<div id="fig-unnamed-chunk-11-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/lukeholman/cost_of_SR_Dpseudo/blob/ffdc5d44370af9cba1f49ab7045fd537d8a2191c/docs/figure/population_genetic_model.Rmd/unnamed-chunk-11-1.png" target="_blank">ffdc5d4</a>
</td>
<td>
lukeholman
</td>
<td>
2019-06-28
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><br></br> <strong>Figure 3</strong>: Comparison of the <em>SR</em> frequencies predicted by the model with the frequencies observed in the wild across seven North American populations. Each point represents one of the populations plotted in Figure 2, and the colour of the point indicates the frequency of female multiple mating in that population. More polyandrous populations contain a lower frequency of <em>SR</em> chromosomes, both in nature and in the model predictions, and the predictions are most accurate when we assume that <em>SR</em> males have similar or equal survival and mating success to <em>ST</em> males (middle and right panel). The dashed line shows <span class="math inline">\(y=x\)</span>, such that plots in which the points are closer to the line indicate a better match between the predicted and observed allele frequencies.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.5.1 (2018-07-02)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.5/Resources/lib/libRlapack.dylib

locale:
[1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] showtext_0.5-1     showtextdb_2.0     sysfonts_0.7.2    
[4] ggplot2_3.1.0      kableExtra_0.9.0   future.apply_1.3.0
[7] future_1.14.0      dplyr_0.8.0.1      purrr_0.3.2       

loaded via a namespace (and not attached):
 [1] tidyselect_0.2.5   xfun_0.8           listenv_0.7.0     
 [4] colorspace_1.3-2   htmltools_0.3.6    viridisLite_0.3.0 
 [7] yaml_2.2.0         rlang_0.4.0        pillar_1.3.1.9000 
[10] glue_1.3.1.9000    withr_2.1.2        RColorBrewer_1.1-2
[13] plyr_1.8.4         stringr_1.4.0      munsell_0.5.0     
[16] gtable_0.2.0       workflowr_1.3.0    rvest_0.3.2       
[19] codetools_0.2-15   evaluate_0.14      labeling_0.3      
[22] knitr_1.23         parallel_3.5.1     curl_3.3          
[25] highr_0.8          Rcpp_1.0.2         readr_1.1.1       
[28] backports_1.1.2    scales_1.0.0       jsonlite_1.6      
[31] fs_1.3.1           hms_0.4.2          digest_0.6.20     
[34] stringi_1.4.3      rprojroot_1.3-2    grid_3.5.1        
[37] tools_3.5.1        magrittr_1.5       lazyeval_0.2.2    
[40] tibble_2.0.99.9000 crayon_1.3.4       whisker_0.3-2     
[43] pkgconfig_2.0.2    xml2_1.2.0         assertthat_0.2.1  
[46] rmarkdown_1.13     httr_1.4.0         rstudioapi_0.10   
[49] R6_2.4.0           globals_0.12.4     git2r_0.23.0      
[52] compiler_3.5.1    </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
